### PR TITLE
Add generate_mask_labels_op to support Mask-RCNN and refine some code.

### PR DESCRIPTION
**Pull request by qingqing01**
1. Add generate_mask_labels_op to support Mask-RCNN.
2. Refine sigmoid_cross_entropy to support nomalize mode.
3. Fix generator_proposals_label.
4. Use DeviceTemporaryAllocator in roi_pool and roi_algin. 
5. Fix https://github.com/PaddlePaddle/models/issues/1649 https://github.com/PaddlePaddle/Paddle/issues/15317

### CODE REVIEW

File: paddle/fluid/operators/sigmoid_cross_entropy_with_logits_op.cu [hunk 1]
```diff
@@ -11,12 +11,184 @@ distributed under the License is distributed on an "AS IS" BASIS,
...
+    auto cnt_ptr = allocator.Allocate(Labels->numel() * sizeof(T));
+    T *counts = reinterpret_cast<T *>(cnt_ptr->ptr());
+
+    int limit = Out->numel();
+    int blocks = NumBlocks(limit);
+    int threads = kNumCUDAThreads;
+    GPUSigmoidForward<T><<<blocks, threads, 0, dev_ctx.stream()>>>(
+        X->data<T>(), Labels->data<T>(), ignore_index, limit, out_data, counts);
+    if (normalize) {
+      auto norm_ptr = allocator.Allocate(sizeof(T));
```
Comment (jerrywgz) [inline paddle/fluid/operators/sigmoid_cross_entropy_with_logits_op.cu:134]: No need for TemporaryAllocator here？

File: python/paddle/fluid/tests/test_detection.py [hunk 2]
```diff
@@ -351,41 +406,43 @@ def test_rpn_target_assign(self):
...
-        self.assertIsNotNone(rpn_rois)
-        self.assertIsNotNone(rpn_roi_probs)
-        print(rpn_rois.shape)
+        program = Program()
+        with program_guard(program):
+            data_shape = [20, 64, 64]
+            images = fluid.layers.data(
+                name='images', shape=data_shape, dtype='float32')
+            im_info = fluid.layers.data(
+                name='im_info', shape=[1, 3], dtype='float32')
```
Comment (jerrywgz) [inline python/paddle/fluid/tests/test_detection.py]: shape should be [3]

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+        ctx->HasOutput("RoiHasMaskInt32"),
+        "Output(RoiHasMaskInt32) of GenerateMaskLabelsOp should not be null");
+    PADDLE_ENFORCE(
+        ctx->HasOutput("MaskInt32"),
+        "Output(MaskInt32) of GenerateMaskLabelsOp should not be null");
+
+    auto im_info_dims = ctx->GetInputDim("ImInfo");
+    //  auto gt_classes_dims = ctx->GetInputDim("GtClasses");
+    //  auto is_crowd_dims = ctx->GetInputDim("IsCrowd");
+    //  auto gt_segms_dims = ctx->GetInputDim("GtSegms");
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/generate_mask_labels_op.cc]: need PADDLE_ENFORCE_EQ(gt_segms_dims[1], 2,"The last dimension of Input(GtSegms) must be 2.")?

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    AddInput("GtClasses",
+             "(LoDTensor), This input is a 2D LoDTensor with shape [M, 1]. "
+             "M is the number of groundtruth, "
+             "each element is a class label of groundtruth.");
+    AddInput(
+        "IsCrowd",
+        "(LoDTensor), This input is a 2D LoDTensor with shape [M, 1]. "
+        "M is the number of groundtruth, "
+        "each element is a flag indicates whether a groundtruth is crowd.");
+    AddInput("GtSegms",
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/generate_mask_labels_op.cc]: 需要修改为poly的格式

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+        }
+      }
+      fg_masks_inds.push_back(id);
+    }
+
+    // add fg targets
+    for (int64_t i = 0; i < fg_num; ++i) {
+      int fg_polys_ind = fg_masks_inds[i];
+      T* roi_fg = rois_fg_data + i * 4;
+      uint8_t* mask = masks_data + i * resolution * resolution;
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/generate_mask_labels_op.cc:223]: comments need to be removed?

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc (hunk 3 above)
Comment (qingqing01) [reply paddle/fluid/operators/detection/generate_mask_labels_op.cc]: Done.

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc (hunk 5 above)
Comment (qingqing01) [reply paddle/fluid/operators/detection/generate_mask_labels_op.cc:223]: Done.

File: paddle/fluid/operators/detection/generate_mask_labels_op.cc (hunk 4 above)
Comment (qingqing01) [reply paddle/fluid/operators/detection/generate_mask_labels_op.cc]: Done.

File: paddle/fluid/operators/sigmoid_cross_entropy_with_logits_op.cu (hunk 1 above)
Comment (qingqing01) [reply paddle/fluid/operators/sigmoid_cross_entropy_with_logits_op.cu:134]: Why not need? It needs here.

File: python/paddle/fluid/tests/test_detection.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/fluid/tests/test_detection.py]: done.

File: python/paddle/fluid/tests/test_detection.py [hunk 6]
```diff
@@ -250,7 +255,7 @@ def test_generate_mask_labels(self):
                 append_batch_size=False)
             gt_segms = layers.data(
                 name='gt_segms',
-                shape=[2, 100, 200],
+                shape=[20, 2],
                 dtype='uint8',
                 lod_level=1,
```
Comment (jerrywgz) [inline python/paddle/fluid/tests/test_detection.py]: lod_level=3

Comment (qingqing01) [reply python/paddle/fluid/tests/test_detection.py]: Done.

**Review by jerrywgz — DISMISSED** [review]

**Review by panyx0718 — APPROVED** [review]
LG API
